# Project: Investigate TMDB Dataset

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
<li><a href="#eda">Exploratory Data Analysis</a></li>
<li><a href="#conclusions">Conclusions</a></li>
</ul>

<a id='intro'></a>
## Introduction

> The dataset that I have selected for investigating is a dataset that contains information about 10,000 movies collected from The Movie Database(TMDb), including user ratings and revenue.
>
## Questions

1. These are the following questions I want to analyze using this dataset:
    - Which genres are receiving higher user ratings?
    - What kinds of properties are associated with movies that have high revenues? 
    - Does higher budget movies have higher user ratings?

In [173]:
# import packages
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as ply

%matplotlib inline
sns.set_style('darkgrid')

<a id='wrangling'></a>
## Data Wrangling

> In this section of the report, I will load in the data, check for cleanliness, and then trim and clean my dataset for analysis.
### General Properties

In [174]:
# load data
file_path = 'tmdb_movies.csv'
movie_df = pd.read_csv(file_path)

In [176]:
# view data
movie_df.head()

id    imdb_id  popularity     budget     revenue  \
0  135397  tt0369610   32.985763  150000000  1513528810   
1   76341  tt1392190   28.419936  150000000   378436354   
2  262500  tt2908446   13.112507  110000000   295238201   
3  140607  tt2488496   11.173104  200000000  2068178225   
4  168259  tt2820852    9.335014  190000000  1506249360   

                 original_title  \
0                Jurassic World   
1            Mad Max: Fury Road   
2                     Insurgent   
3  Star Wars: The Force Awakens   
4                     Furious 7   

                                                cast  \
0  Chris Pratt|Bryce Dallas Howard|Irrfan Khan|Vi...   
1  Tom Hardy|Charlize Theron|Hugh Keays-Byrne|Nic...   
2  Shailene Woodley|Theo James|Kate Winslet|Ansel...   
3  Harrison Ford|Mark Hamill|Carrie Fisher|Adam D...   
4  Vin Diesel|Paul Walker|Jason Statham|Michelle ...   

                                            homepage          director  \
0                      http://www.jurassicworld.com/   Colin Trevorrow   
1                        http://www.madmaxmovie.com/     George Miller   
2     http://www.thedivergentseries.movie/#insurgent  Robert Schwentke   
3  http://www.starwars.com/films/star-wars-episod...       J.J. Abrams   
4                           http://www.furious7.com/         James Wan   

                         tagline  ...  \
0              The park is open.  ...   
1             What a Lovely Day.  ...   
2     One Choice Can Destroy You  ...   
3  Every generation has a story.  ...   
4            Vengeance Hits Home  ...   

                                            overview runtime  \
0  Twenty-two years after the events of Jurassic ...     124   
1  An apocalyptic story set in the furthest reach...     120   
2  Beatrice Prior must confront her inner demons ...     119   
3  Thirty years after defeating the Galactic Empi...     136   
4  Deckard Shaw seeks revenge against Dominic Tor...     137   

                                      genres  \
0  Action|Adventure|Science Fiction|Thriller   
1  Action|Adventure|Science Fiction|Thriller   
2         Adventure|Science Fiction|Thriller   
3   Action|Adventure|Science Fiction|Fantasy   
4                      Action|Crime|Thriller   

                                production_companies release_date vote_count  \
0  Universal Studios|Amblin Entertainment|Legenda...       6/9/15       5562   
1  Village Roadshow Pictures|Kennedy Miller Produ...      5/13/15       6185   
2  Summit Entertainment|Mandeville Films|Red Wago...      3/18/15       2480   
3          Lucasfilm|Truenorth Productions|Bad Robot     12/15/15       5292   
4  Universal Pictures|Original Film|Media Rights ...       4/1/15       2947   

   vote_average  release_year    budget_adj   revenue_adj  
0           6.5          2015  1.379999e+08  1.392446e+09  
1           7.1          2015  1.379999e+08  3.481613e+08  
2           6.3          2015  1.012000e+08  2.716190e+08  
3           7.5          2015  1.839999e+08  1.902723e+09  
4           7.3          2015  1.747999e+08  1.385749e+09  

[5 rows x 21 columns]

In [177]:
# shape of the data
print('Number of Movies: {}'.format(movie_df.shape[0]))
print('Number of Movie Attributes: {}'.format(movie_df.shape[1]))

Number of Movies: 10866
Number of Movie Attributes: 21


In [178]:
# number of nulls in each column
movie_df.isnull().sum()

id                         0
imdb_id                   10
popularity                 0
budget                     0
revenue                    0
original_title             0
cast                      76
homepage                7930
director                  44
tagline                 2824
keywords                1493
overview                   4
runtime                    0
genres                    23
production_companies    1030
release_date               0
vote_count                 0
vote_average               0
release_year               0
budget_adj                 0
revenue_adj                0
dtype: int64

### Observation 1
After investigating dataset for number of nulls in each column, this is what I found:
- **homepage** has `7930` nulls
- **tagline** has `2824` nulls
- **keywords** has `1493` nulls
- **overview** has `4` nulls
- **imdb_id** has `10` nulls
- **release_date**

There are also other columns with nulls but I decided to drop these columns from my dataframe as there are lot of nulls and if you observe there are some columns with less number of nulls that I'm gonna drop as these columns are not much useful for my analysis.

And talking about the remaining nulls I'm gonna handle them in my next steps.

In [179]:
# drop homepage, tagline, keywords, overview, and imdb_id columns
drop_columns = ['imdb_id', 'homepage', 'tagline', 'keywords', 'overview', 'release_date']
movie_df.drop(columns = drop_columns, inplace = True)

In [180]:
# verify with shape
print('Number of Movies: {}'.format(movie_df.shape[0]))
print('Number of Movie Attributes: {}'.format(movie_df.shape[1]))

Number of Movies: 10866
Number of Movie Attributes: 15


In [181]:
# finding duplicates
movie_df[movie_df.duplicated()]

id  popularity    budget  revenue original_title  \
2090  42194     0.59643  30000000   967000         TEKKEN   

                                                   cast          director  \
2090  Jon Foo|Kelly Overton|Cary-Hiroyuki Tagawa|Ian...  Dwight H. Little   

      runtime                                       genres  \
2090       92  Crime|Drama|Action|Thriller|Science Fiction   

        production_companies  vote_count  vote_average  release_year  \
2090  Namco|Light Song Films         110           5.0          2010   

      budget_adj  revenue_adj  
2090  30000000.0     967000.0

In [182]:
movie_df.drop_duplicates(inplace = True)

In [183]:
# verify duplicates
print('Your data contains {} duplicates'.format(movie_df.duplicated().sum()))

Your data contains 0 duplicates


### Observation 2
After inspecting for duplicates in my dataframe there is only one duplicate and I dropped it from my dataframe.

In [184]:
# finding rows with missing values
movie_df[movie_df.isnull().any(axis = 1)]

id  popularity  budget  revenue  \
228    300792    0.584363       0        0   
259    360603    0.476341       0        0   
295    363483    0.417191       0        0   
298    354220    0.370258       0        0   
328    308457    0.367617       0        0   
...       ...         ...     ...      ...   
10804   15867    0.149259       0        0   
10806   24998    0.138635       0        0   
10816   16378    0.064602       0        0   
10842   36540    0.253437       0        0   
10853   15598    0.163592       0        0   

                           original_title  \
228                     Racing Extinction   
259                   Crown for Christmas   
295                 12 Gifts of Christmas   
298           The Girl in the Photographs   
328                          Advantageous   
...                                   ...   
10804                           Interiors   
10806                     Gates of Heaven   
10816    The Rutles: All You Need Is Cash   
10842  Winnie the Pooh and the Honey Tree   
10853                               Alfie   

                                                    cast             director  \
228    Elon Musk|Jane Goodall|Louie Psihoyos|Leilani ...       Louie Psihoyos   
259    Danica McKellar|Rupert Penry-Jones|Ellie Botte...            Alex Zamm   
295    Katrina Law|Donna Mills|Aaron O'Connell|Melani...       Peter Sullivan   
298    Kal Penn|Claudia Lee|Kenny Wormald|Toby Heming...           Nick Simon   
328    Jacqueline Kim|James Urbaniak|Freya Adams|Ken ...       Jennifer Phang   
...                                                  ...                  ...   
10804  Diane Keaton|Kristin Griffith|Mary Beth Hurt|R...          Woody Allen   
10806  Lucille Billingsley|Zella Graham|Cal Harberts|...         Errol Morris   
10816  Eric Idle|John Halsey|Ricky Fataar|Neil Innes|...  Eric Idle|Gary Weis   
10842  Sterling Holloway|Junius Matthews|Sebastian Ca...  Wolfgang Reitherman   
10853  Michael Caine|Shelley Winters|Millicent Martin...        Lewis Gilbert   

       runtime                        genres production_companies  vote_count  \
228         90         Adventure|Documentary                  NaN          36   
259         84                      TV Movie                  NaN          10   
295         84               Family|TV Movie                  NaN          12   
298         95         Crime|Horror|Thriller                  NaN          10   
328         92  Science Fiction|Drama|Family                  NaN          29   
...        ...                           ...                  ...         ...   
10804       93                         Drama                  NaN          35   
10806       85                   Documentary                  NaN          12   
10816       76                        Comedy                  NaN          14   
10842       25              Animation|Family                  NaN          12   
10853      114          Comedy|Drama|Romance                  NaN          26   

       vote_average  release_year  budget_adj  revenue_adj  
228             7.8          2015         0.0          0.0  
259             7.6          2015         0.0          0.0  
295             6.3          2015         0.0          0.0  
298             4.7          2015         0.0          0.0  
328             6.4          2015         0.0          0.0  
...             ...           ...         ...          ...  
10804           6.3          1978         0.0          0.0  
10806           5.9          1978         0.0          0.0  
10816           6.0          1978         0.0          0.0  
10842           7.9          1966         0.0          0.0  
10853           6.2          1966         0.0          0.0  

[1093 rows x 15 columns]

### Observation 3
After inspecting the number of rows with missing values I have decided to drop them my dataframe.

In [185]:
# drop rows with missing columns
movie_df.drop(index = movie_df[movie_df.isnull().any(axis = 1)].index, inplace = True)

In [186]:
# verfiy for missing values
print('Your data contains {} missing values'.format(movie_df.isnull().sum().sum()))

Your data contains 0 missing values


In [187]:
# inspect data types
movie_df.dtypes

id                        int64
popularity              float64
budget                    int64
revenue                   int64
original_title           object
cast                     object
director                 object
runtime                   int64
genres                   object
production_companies     object
vote_count                int64
vote_average            float64
release_year              int64
budget_adj              float64
revenue_adj             float64
dtype: object

### Observation 4
After inspecting for data types I decided to perform following things:
- Convert `budget`, `revenue` and `runtime` columns to floats

In [188]:
# Convert budget, revenue and runtime columns to floats
movie_df[['budget', 'revenue']] = movie_df[['budget', 'revenue']].astype(float)

In [189]:
# verify data types
movie_df.dtypes

id                        int64
popularity              float64
budget                  float64
revenue                 float64
original_title           object
cast                     object
director                 object
runtime                   int64
genres                   object
production_companies     object
vote_count                int64
vote_average            float64
release_year              int64
budget_adj              float64
revenue_adj             float64
dtype: object

In [190]:
# Checking for consistency
movie_df.describe()

id   popularity        budget       revenue      runtime  \
count    9772.000000  9772.000000  9.772000e+03  9.772000e+03  9772.000000   
mean    63189.640810     0.694721  1.617967e+07  4.423121e+07   102.926627   
std     90718.059987     1.036931  3.221074e+07  1.225889e+08    27.877432   
min         5.000000     0.000188  0.000000e+00  0.000000e+00     0.000000   
25%     10221.500000     0.232710  0.000000e+00  0.000000e+00    90.000000   
50%     18677.500000     0.419762  2.000000e+05  0.000000e+00   100.000000   
75%     70577.250000     0.776408  1.928750e+07  3.104729e+07   112.000000   
max    417859.000000    32.985763  4.250000e+08  2.781506e+09   877.000000   

        vote_count  vote_average  release_year    budget_adj   revenue_adj  
count  9772.000000   9772.000000   9772.000000  9.772000e+03  9.772000e+03  
mean    239.312014      5.963528   2000.878428  1.941599e+07  5.705309e+07  
std     603.011504      0.913174     13.036794  3.566634e+07  1.514499e+08  
min      10.000000      1.500000   1960.000000  0.000000e+00  0.000000e+00  
25%      18.000000      5.400000   1994.000000  0.000000e+00  0.000000e+00  
50%      46.000000      6.000000   2005.000000  3.061342e+05  0.000000e+00  
75%     173.000000      6.600000   2011.000000  2.464268e+07  4.311848e+07  
max    9767.000000      8.700000   2015.000000  4.250000e+08  2.827124e+09

In [191]:
movie_df.head()

id  popularity       budget       revenue  \
0  135397   32.985763  150000000.0  1.513529e+09   
1   76341   28.419936  150000000.0  3.784364e+08   
2  262500   13.112507  110000000.0  2.952382e+08   
3  140607   11.173104  200000000.0  2.068178e+09   
4  168259    9.335014  190000000.0  1.506249e+09   

                 original_title  \
0                Jurassic World   
1            Mad Max: Fury Road   
2                     Insurgent   
3  Star Wars: The Force Awakens   
4                     Furious 7   

                                                cast          director  \
0  Chris Pratt|Bryce Dallas Howard|Irrfan Khan|Vi...   Colin Trevorrow   
1  Tom Hardy|Charlize Theron|Hugh Keays-Byrne|Nic...     George Miller   
2  Shailene Woodley|Theo James|Kate Winslet|Ansel...  Robert Schwentke   
3  Harrison Ford|Mark Hamill|Carrie Fisher|Adam D...       J.J. Abrams   
4  Vin Diesel|Paul Walker|Jason Statham|Michelle ...         James Wan   

   runtime                                     genres  \
0      124  Action|Adventure|Science Fiction|Thriller   
1      120  Action|Adventure|Science Fiction|Thriller   
2      119         Adventure|Science Fiction|Thriller   
3      136   Action|Adventure|Science Fiction|Fantasy   
4      137                      Action|Crime|Thriller   

                                production_companies  vote_count  \
0  Universal Studios|Amblin Entertainment|Legenda...        5562   
1  Village Roadshow Pictures|Kennedy Miller Produ...        6185   
2  Summit Entertainment|Mandeville Films|Red Wago...        2480   
3          Lucasfilm|Truenorth Productions|Bad Robot        5292   
4  Universal Pictures|Original Film|Media Rights ...        2947   

   vote_average  release_year    budget_adj   revenue_adj  
0           6.5          2015  1.379999e+08  1.392446e+09  
1           7.1          2015  1.379999e+08  3.481613e+08  
2           6.3          2015  1.012000e+08  2.716190e+08  
3           7.5          2015  1.839999e+08  1.902723e+09  
4           7.3          2015  1.747999e+08  1.385749e+09